Test/civl/async/async4.bpl

```diff
@@ -0,0 +1,33 @@
+// RUN: %boogie "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+var {:layer 0,2} x:int;
+
+type {:pending_async}{:datatype} PA;
+function {:constructor} A_Inc() : PA;
+
+procedure {:both}{:layer 2} A_Add (n: int)
+modifies x;
+{ assert 0 <= n; x := x + n; }
+
+procedure {:IS_invariant}{:layer 1} INV(n: int) returns ({:pending_async "A_Inc"} PAs: [PA]int)
+modifies x;
+{
+  var {:inst_at "A"} i: int;
+  assert 0 <= n;
+  assume {:inst "A", i} {:inst "A", i+1} 0 <= i && i <= n;
+  x := x + i;
+  PAs := MapConst(0)[A_Inc() := n - i];
```

Comment:
@bkragl :  If I write this assignment as 

PAs[A_Inc()] := n - i

it is logically equivalent but the verification does not work. This happens because the initial value of the output variable PAs becomes an existentially quantified variable of the transition relation.  I could have supplied a hint for this variable also but that would have been no better than the alternative I chose.

Reply:
Yes, I see. But this is a special case where `PAs[A_Inc()] := n - i` only works because the domain of `PAs` is a singleton (because `PA` has only a single constructor with no parameters). I think the way you write it is much better than supplying a hint.

Reply:
Did you notice the syntax highlighting we have now for Boogie code? :smiley:

Reply:
I hadn't noticed it.  But I see it now.  Cool!